exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/project/ProjectMemoryManager.java

```diff
@@ -0,0 +1,315 @@
+        }
+        outputColumnSizes.put(columnWidthInfo.getName(), columnWidthInfo);
+    }
+
+    void addComplexField(ValueVector vv) {
+        //Complex types are not yet supported. Just use a guess for the size
+        assert vv == null || isComplex(vv.getField().getType());
+        complexColumnsCount++;
+        // just a guess
+        totalComplexColumnWidth +=  OutputSizeEstimateConstants.COMPLEX_FIELD_ESTIMATE;
+    }
+
+    void addFixedWidthField(ValueVector vv) {
+        assert isFixedWidth(vv);
+        fixedWidthColumnCount++;
+        int fixedFieldWidth = getWidthOfFixedWidthType(vv);
+        totalFixedWidthColumnWidth += fixedFieldWidth;
+    }
+
+    public void init(RecordBatch incomingBatch, ProjectRecordBatch outgoingBatch) {
+        setIncomingBatch(incomingBatch);
+        setOutgoingBatch(outgoingBatch);
+        reset();
+    }
+
+    private void reset() {
+        rowWidth = 0;
+        totalFixedWidthColumnWidth = 0;
+        totalComplexColumnWidth = 0;
+
+        fixedWidthColumnCount = 0;
+        complexColumnsCount = 0;
+    }
+
+    @Override
+    public void update() {
+        long updateStartTime = System.currentTimeMillis();
+        RecordBatchSizer batchSizer = new RecordBatchSizer(incomingBatch);
+        long batchSizerEndTime = System.currentTimeMillis();
+
+        setRecordBatchSizer(batchSizer);
+        rowWidth = 0;
+        int totalVariableColumnWidth = 0;
+        for (String expr : outputColumnSizes.keySet()) {
+            ColumnWidthInfo columnWidthInfo = outputColumnSizes.get(expr);
+            int width = -1;
+            if (columnWidthInfo.isFixedWidth()) {
+                // fixed width columns are accumulated in totalFixedWidthColumnWidth
+                ShouldNotReachHere();
+            } else {
+                //Walk the tree of OutputWidthExpressions to get a FixedLenExpr
+                //As the tree is walked, the RecordBatchSizer and function annotations
+                //are looked-up to come up with the final FixedLenExpr
+                OutputWidthExpression savedWidthExpr = columnWidthInfo.getOutputExpression();
+                OutputColumnType columnType = columnWidthInfo.getOutputColumnType();
+                OutputWidthVisitorState state = new OutputWidthVisitorState(this, columnType);
+                OutputWidthExpression reducedExpr = savedWidthExpr.accept(new OutputWidthVisitor(), state);
+                assert reducedExpr instanceof FixedLenExpr;
+                width = ((FixedLenExpr)reducedExpr).getWidth();
+                assert width >= 0;
+            }
+            totalVariableColumnWidth += width;
+        }
+        rowWidth += totalFixedWidthColumnWidth;
+        rowWidth += totalComplexColumnWidth;
+        rowWidth += totalVariableColumnWidth;
+        int outPutRowCount;
+        if (rowWidth != 0) {
+            setOutputRowCount(getOutputBatchSize(), rowWidth);
```

Comment:
Maybe add a little comment here; this is the important place in the code where the output size for Project is being set ....